File: Embarcadero/XE5/ObjectPascal/LocationDemo/LocationDemoUnit.pas
unit LocationDemoUnit;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Dialogs, FMX.Objects, FMX.StdCtrls,
  FMX.ListBox, FMX.Layouts, FMX.WebBrowser, FMX.Sensors, System.Sensors, FMX.MobilePreview;

type
  TForm1 = class(TForm)
    LocationSensor1: TLocationSensor;
    WebBrowser1: TWebBrowser;
    ListBox1: TListBox;
    ListBoxItem1: TListBoxItem;
    Switch1: TSwitch;
    ListBoxGroupHeader1: TListBoxGroupHeader;
    ListBoxItemLatitude: TListBoxItem;
    ListBoxItemLongitude: TListBoxItem;
    ListBoxGroupHeader2: TListBoxGroupHeader;
    ListBoxItemAdminArea: TListBoxItem;
    ListBoxItemCountryCode: TListBoxItem;
    ListBoxItemCountryName: TListBoxItem;
    ListBoxItemFeatureName: TListBoxItem;
    ListBoxItemLocality: TListBoxItem;
    ListBoxItemPostalCode: TListBoxItem;
    ListBoxItemSubAdminArea: TListBoxItem;
    ListBoxItemSubLocality: TListBoxItem;
    ListBoxItemSubThoroughfare: TListBoxItem;
    ListBoxItemThoroughfare: TListBoxItem;
    ListBoxHeader1: TListBoxHeader;
    Label1: TLabel;
    Line2: TLine;
    procedure LocationSensor1LocationChanged(Sender: TObject; const OldLocation,
      NewLocation: TLocationCoord2D);
    procedure Switch1Switch(Sender: TObject);
  private
    { Private declarations }
    FGeocoder: TGeocoder;
    procedure OnGeocodeReverseEvent(const Address: TCivicAddress);
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.fmx}
procedure TForm1.Switch1Switch(Sender: TObject);
begin
  LocationSensor1.Active := Switch1.IsChecked;
end;

procedure TForm1.LocationSensor1LocationChanged(Sender: TObject;
  const OldLocation, NewLocation: TLocationCoord2D);
var
  URLString: String;
  LLatitude, LLongitude : string;
  LSettings: TFormatSettings;
  LDecSeparator : Char;
begin
  LDecSeparator := FormatSettings.DecimalSeparator;
  LSettings := FormatSettings;
  try
    FormatSettings.DecimalSeparator := '.';
    // Show current location
    ListBoxItemLatitude.ItemData.Detail  := Format('%2.6f', [NewLocation.Latitude]);
    ListBoxItemLongitude.ItemData.Detail := Format('%2.6f', [NewLocation.Longitude]);

    // Show Map using Google Maps
    LLongitude := FloatToStr(NewLocation.Longitude, LSettings);
    URLString := Format(
      'https://maps.google.com/maps?q=%2.6f,%2.6f&output=embed',
        [ NewLocation.Latitude, NewLocation.Longitude]);
  finally
    FormatSettings.DecimalSeparator := LDecSeparator;
  end;
  WebBrowser1.Navigate(URLString);

  // Setup an instance of TGeocoder
  if not Assigned(FGeocoder) then
  begin
    if Assigned(TGeocoder.Current) then
      FGeocoder := TGeocoder.Current.Create;
    if Assigned(FGeocoder) then
      FGeocoder.OnGeocodeReverse := OnGeocodeReverseEvent;
  end;

  // Translate location to address
  if Assigned(FGeocoder) and not FGeocoder.Geocoding then
    FGeocoder.GeocodeReverse(NewLocation);
end;

procedure TForm1.OnGeocodeReverseEvent(const Address: TCivicAddress);
begin
  if Address <> nil then
  begin
    ListBoxItemAdminArea.ItemData.Detail       := Address.AdminArea;
    ListBoxItemCountryCode.ItemData.Detail     := Address.CountryCode;
    ListBoxItemCountryName.ItemData.Detail     := Address.CountryName;
    ListBoxItemFeatureName.ItemData.Detail     := Address.FeatureName;
    ListBoxItemLocality.ItemData.Detail        := Address.Locality;
    ListBoxItemPostalCode.ItemData.Detail      := Address.PostalCode;
    ListBoxItemSubAdminArea.ItemData.Detail    := Address.SubAdminArea;
    ListBoxItemSubLocality.ItemData.Detail     := Address.SubLocality;
    ListBoxItemSubThoroughfare.ItemData.Detail := Address.SubThoroughfare;
    ListBoxItemThoroughfare.ItemData.Detail    := Address.Thoroughfare;
  end;
end;
end.

File: Embarcadero/XE5/CPP/LocationDemo/LocationDemoUnit.cpp
//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "LocationDemoUnit.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TForm2 *Form2;
//---------------------------------------------------------------------------
__fastcall TForm2::TForm2(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm2::Switch1Switch(TObject *Sender)
{
	LocationSensor1->Active = Switch1->IsChecked;
}
//---------------------------------------------------------------------------
void __fastcall TForm2::OnGeocodeReverseEvent(TCivicAddress* const Address)
{
	if (Address != NULL){
		ListBoxItemAdminArea->ItemData->Detail       = Address->AdminArea;
		ListBoxItemCountryCode->ItemData->Detail     = Address->CountryCode;
		ListBoxItemCountryName->ItemData->Detail     = Address->CountryName;
		ListBoxItemFeatureName->ItemData->Detail     = Address->FeatureName;
		ListBoxItemLocality->ItemData->Detail        = Address->Locality;
		ListBoxItemPostalCode->ItemData->Detail      = Address->PostalCode;
		ListBoxItemSubAdminArea->ItemData->Detail    = Address->SubAdminArea;
		ListBoxItemSubLocality->ItemData->Detail     = Address->SubLocality;
		ListBoxItemSubThoroughfare->ItemData->Detail = Address->SubThoroughfare;
		ListBoxItemThoroughfare->ItemData->Detail    = Address->Thoroughfare;
	}
}
//---------------------------------------------------------------------------


void __fastcall TForm2::LocationSensor1LocationChanged(TObject *Sender, const TLocationCoord2D &OldLocation,
          const TLocationCoord2D &NewLocation)
{
	WideChar LDecSeparator = FormatSettings.DecimalSeparator;
	TFormatSettings LSettings = FormatSettings;
	FormatSettings.DecimalSeparator = '.';
	// Show current location
	ListBoxItemLatitude->ItemData->Detail =
		ListBoxItemLatitude->ItemData->Detail.sprintf(L"%2.6f", NewLocation.Latitude);
	ListBoxItemLongitude->ItemData->Detail =
		ListBoxItemLongitude->ItemData->Detail.sprintf(L"%2.6f", NewLocation.Longitude);
	// Show Map using Google Maps
	auto LLongitude = FloatToStr(NewLocation.Longitude, LSettings);
	String URLString = "";
	URLString = URLString.sprintf(L"https://maps.google.com/maps?q=%2.6f,%2.6f&output=embed",
		NewLocation.Latitude, NewLocation.Longitude);

	FormatSettings.DecimalSeparator = LDecSeparator;
	WebBrowser1->Navigate(URLString);

	// Setup an instance of TGeocoder
	if (FGeocoder == NULL) {
		if (TGeocoder::Current != NULL) {
			FGeocoder = (TGeocoder*)new TGeocoderClass(TGeocoder::Current);
		}
		if (FGeocoder != NULL) {
			FGeocoder->OnGeocodeReverse = OnGeocodeReverseEvent;
		}
	}
	// Translate location to address
	if ((FGeocoder != NULL) && (FGeocoder->Geocoding)) {
		FGeocoder->GeocodeReverse(NewLocation);
	}
}
//---------------------------------------------------------------------------

